Dockerfile

```diff
@@ -0,0 +1,24 @@
+FROM alpine as prezto-devel
+LABEL maintainer="h@hlo.mx"
+RUN apk --no-cache update && apk upgrade && \
+    apk add util-linux pciutils usbutils coreutils binutils\
+    findutils grep man man-pages mdocml-apropos less less-doc \
+    make grep zsh zsh-vcs zsh-zftp zsh-calendar zsh-doc git \
+    vim git-zsh-completion tmux tmux-zsh-completion tree \
+    docker-zsh-completion
+RUN addgroup prezto && adduser -D prezto -G prezto -S /bin/zsh
+WORKDIR /home/prezto
+USER prezto
+RUN cd /home/prezto && mkdir src
+COPY . src
```

Comment:
If you used `ADD` instead of `COPY` it would speed up things, because then the files would not really be copied into the container.  
Or do you need it like that so that your local sources won't be changed with the later make runs?

Reply:
Can you please tell me how `ADD` will speed things up? According to docker's [best practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy) individual `COPY` statements are preferred.

Reply:
ADD adds the directory as a docker volume.  
COPY really creates copies of the files into the docker container.

At least that was how I understood it so far...

Reply:
I don't think it works like that. Neither statements add a directory as a volume. Docker volumes are different directives manipulated by VOLUME statement in Dockerfile or `-v` flag at runtime.

Both `ADD` and `COPY` statements copy the source from the build context (first argument) to the destination in docker image (second argument).

Specifically, as the best practices link explains, the ADD statement has additional magic to it - it will _automagically_ uncompress a compressed file, or if the first argument is a URL it will download it.

`COPY . src` is good first pass for a Dockerfile. Though ideally, one would prefer to have separate statements for each directory, so that a change in one file in one directory will not invalidate the entire cache. But then we run into the problem of how we will order individual COPY statements. So, I am okay with `COPY . src` for now.

Reply:
Well however, it doesn't really matter to me actually if ADD or COPY is used in the end here. 

My main point is not running stuff in the image building process, but rather run it on the finished image... 
Its just quicker, and sllicker. If you prefer to build the image newly each time. you run something, or change code and then test it, and fail and change it again... up to you. :wink: